javascript/ql/src/semmle/javascript/Collections.qll

```diff
@@ -0,0 +1,279 @@
+    DataFlow::Node pred, DataFlow::Node succ, string loadProp, string storeProp
+  ) {
+    this.loadStore(pred, succ, loadProp, storeProp)
+  }
+}
+
+/**
+ * A collection of predicates and clases for type-tracking collections.
+ */
+module CollectionsTypeTracking {
+  /**
+   * Gets the result from a single step through a collection, from `pred` to `result` summarized by `summary`.
+   */
+  pragma[inline]
+  DataFlow::SourceNode collectionStep(DataFlow::Node pred, StepSummary summary) {
+    exists(CollectionFlowStep step, PseudoProperty field |
+      summary = LoadStep(field) and
+      step.load(pred, result, field) and
+      not field = mapValueUnknownKey() // prune unknown reads in type-tracking
+      or
+      summary = StoreStep(field) and
+      step.store(pred, result, field)
+      or
+      summary = CopyStep(field) and
+      step.loadStore(pred, result, field)
+      or
+      exists(PseudoProperty toField | summary = LoadStoreStep(field, toField) |
+        step.loadStore(pred, result, field, toField)
+      )
+    )
+  }
+
+  /**
+   * Gets the result from a single step through a collection, from `pred` with tracker `t2` to `result` with tracker `t`.
+   */
+  pragma[inline]
+  DataFlow::SourceNode collectionStep(
+    DataFlow::SourceNode pred, DataFlow::TypeTracker t, DataFlow::TypeTracker t2
+  ) {
+    exists(DataFlow::Node mid, StepSummary summary | pred.flowsTo(mid) and t = t2.append(summary) |
+      result = collectionStep(mid, summary)
+    )
+  }
+}
+
+/**
+ * A module for data-flow steps related standard library collection implementations.
+ */
+private module CollectionDataFlow {
+  /**
+   * A step for `Set.add()` method, which adds an element to a Set.
+   */
+  private class SetAdd extends CollectionFlowStep, DataFlow::MethodCallNode {
+    SetAdd() { this.getMethodName() = "add" }
+
+    override predicate store(DataFlow::Node element, DataFlow::Node obj, PseudoProperty prop) {
+      obj = this.getReceiver().getALocalSource() and
+      element = this.getArgument(0) and
+      prop = setElement()
+    }
+  }
+
+  /**
+   * A step for the `Set` constructor, which copies any elements from the first argument into the resulting set.
+   */
+  private class SetConstructor extends CollectionFlowStep, DataFlow::NewNode {
+    SetConstructor() { this = DataFlow::globalVarRef("Set").getAnInstantiation() }
+
+    override predicate loadStore(
+      DataFlow::Node pred, DataFlow::Node succ, PseudoProperty fromProp, PseudoProperty toProp
+    ) {
+      pred = this.getArgument(0) and
+      succ = this and
+      fromProp = arrayLikeElement() and
+      toProp = setElement()
+    }
+  }
+
+  /**
+   * A step for a `for of` statement on a Map, Set, or Iterator.
+   * For Sets and iterators the l-value are the elements of the set/iterator.
+   * For Maps the l-value is a tuple containing a key and a value.
+   *
+   * This is partially duplicated behavior with the `for of` step for Arrays (in Arrays.qll).
```

Comment:
The `Arrays.qll` file currently has no public members relevant to pseudo-properties.  
Should I still refer to private modules? 

(I could introduce a type-tracking predicate for Arrays. That would both prepare type-tracking predicates for if we need them, and create public members I can refer to).

Reply:
I think it is fine to reference private predicates when you are already discussing an implementation detail like duplication. Perhaps it would be better if this line was moved to a `//`-comment instead to avoid the docstring inclusion though.